Add Product Card component

simranvijwani opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
- Add a new product card component to the new marketplace react page (under WooCommerce WP Admin). 

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/Automattic/woocommerce.com/issues/17431.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch. 
2. Build WooCommerce Admin on your local environment: `cd plugins/woocommerce-admin` then `nvm use && pnpm start`. (If you get a fatal error, refresh the Composer autoload: `cd plugins/woocommerce` and `composer dump-autoload`.)
3. Open `plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx` file. 
4. Add the products using the sample data present at the bottom inside the `woocommerce-marketplace__product-list-content` div. 
5. Visit WooCommerce > Marketplace > Extensions (eg. http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions&tab=extensions)
6. Confirm product cards getting displayed correctly. 

#### Product card sample data
```
<ProductCard
	product={ {
		title: 'Inventory Management for WooCommerce',
		description:
			'Optimize your inventory, predict customer demand, automate stock replenishment, and manage purchase orders with the Inventory Management for WooCommerce extension.',
		vendorName: 'Shelf Planner',
		vendorUrl: 'https://woocommerce.com/payments/',
		icon: 'https://woocommerce.com/wp-content/uploads/2022/10/Favicon-Shelf-Planner-1.png?w=80',
		url: 'https://woocommerce.com/payments/',
		price: 0,
	} }
/>
<ProductCard
	product={ {
		title: 'Inventory Management for WooCommerce',
		description:
			'Optimize your inventory, predict customer demand, automate stock replenishment, and manage purchase orders with the Inventory Management for WooCommerce extension.',
		vendorName: 'Shelf Planner',
		vendorUrl: 'https://woocommerce.com/payments/',
		icon: 'https://woocommerce.com/wp-content/uploads/2022/10/Favicon-Shelf-Planner-1.png?w=80',
		url: 'https://woocommerce.com/payments/',
		price: 0,
	} }
/>
<ProductCard
	product={ {
		title: 'Inventory Management for WooCommerce',
		description:
			'Optimize your inventory, predict customer demand, automate stock replenishment, and manage purchase orders with the Inventory Management for WooCommerce extension.',
		vendorName: 'Shelf Planner',
		vendorUrl: 'https://woocommerce.com/payments/',
		icon: 'https://woocommerce.com/wp-content/uploads/2022/10/Favicon-Shelf-Planner-1.png?w=80',
		url: 'https://woocommerce.com/payments/',
		price: 0,
	} }
/>
<ProductCard
	product={ {
		title: 'Inventory Management for WooCommerce',
		description:
			'Optimize your inventory, predict customer demand, automate stock replenishment, and manage purchase orders with the Inventory Management for WooCommerce extension.',
		vendorName: 'Shelf Planner',
		vendorUrl: 'https://woocommerce.com/payments/',
		icon: 'https://woocommerce.com/wp-content/uploads/2022/10/Favicon-Shelf-Planner-1.png?w=80',
		url: 'https://woocommerce.com/payments/',
		price: 0,
	} }
/>
<ProductCard
	product={ {
		title: 'Inventory Management for WooCommerce',
		description:
			'Optimize your inventory, predict customer demand, automate stock replenishment, and manage purchase orders with the Inventory Management for WooCommerce extension.',
		vendorName: 'Shelf Planner',
		vendorUrl: 'https://woocommerce.com/payments/',
		icon: 'https://woocommerce.com/wp-content/uploads/2022/10/Favicon-Shelf-Planner-1.png?w=80',
		url: 'https://woocommerce.com/payments/',
		price: 0,
	} }
/>
			
```

<img width="1648" alt="image" src="https://github.com/woocommerce/woocommerce/assets/34896144/6fdb60c2-1e6e-4e1c-b7f7-b15013b67371">


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx:
```diff
@@ -1,19 +1,50 @@
+	productType?: string;
+	averageRating?: number | null;
+	reviewsCount?: number | null;
+	currency?: string;
```
(My understanding is that) the `?` makes currency an _optional_ prop. Is that correct?

simranvijwani replied:
Yes, we are not using the currency for now. We might modify the API to include the currency symbol. So this shall be updated later.

Dan-Q reviewed (changes requested):
I get layout problems at some screen widths (tested using Firefox) -

**Between 782 and 1200 pixels wide**: "Long" product titles or their developers wrap and intersect with the product description:

![image](https://github.com/woocommerce/woocommerce/assets/53293/e57582e7-6cf3-4e5e-8e9d-bbd4eb3483c1)

**Between 500 and 780 pixels wide**: "Short" product descriptions result in that product's icon no-longer being left-aligned: 

![image](https://github.com/woocommerce/woocommerce/assets/53293/4b7ed886-0395-4b05-9e3d-c9973b8985a3)

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
_Love_ your use of the base WooCommerce styling for spacing! I'm going to update our `_variables.scss` file to use it.

kdevnel reviewed (changes requested):
Thanks for this Simran! It's starting to look awesome!

1. Similar to a point Dan made, the vendor name on longer vendor names is causing some spacing issues with the paragraph below:
<img width="478" alt="Screenshot 2023-07-21 at 10 08 09" src="https://github.com/woocommerce/woocommerce/assets/22053773/4559d844-12a5-4064-8ad9-cacb673f6457">

---

2. I noticed on my side that one of the cards didn't manage to load a vendor image. It only showed some text in it's place. I wonder if we should have a local placeholder that loads instead - it would look nicer and I can imagine some user sites will have weird connection timeouts that we can't always control. The joys of planning for core!
<img width="505" alt="Screenshot 2023-07-21 at 10 10 11" src="https://github.com/woocommerce/woocommerce/assets/22053773/44121851-11a9-4975-b46e-fb920096bc7e">

---

3. I also noticed that 2 line descriptions lead to the pricing text not appearing at the bottom of the card. I notice the design forces all descriptions to 2 lines by concatenating the text. Not sure how easy that would be to implement, but it would make the cards more consistent with the design. Maybe something to discuss with Poli.
<img width="470" alt="Screenshot 2023-07-21 at 10 21 35" src="https://github.com/woocommerce/woocommerce/assets/22053773/4aa4d285-ccaf-4d06-8f23-c17db96aef05">

---

4. Last thing, I notice the `Card` component you're using here doesn't quite match the design. It looks like it includes a border-radius of 1px instead of the 4px defined in the design.

andfinally commented:
Where we can't find an app icon for a product, I'd vote for just omitting the icon. This is what we do in WCCOM.

<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/859de60d-03a4-4a33-97d6-863028a1886c">

simranvijwani commented:
@andfinally @kdevnel Thank you for the initial review. I'm taking care of the issue [Discover Page: skeleton by parsing as-is Featured API into component with section components](https://github.com/Automattic/woocommerce.com/issues/17433) in this PR itself. I'll re-request your review when the PR is ready. Thanks!

kdevnel commented:
> @andfinally @kdevnel Thank you for the initial review. I'm taking care of the issue [Discover Page: skeleton by parsing as-is Featured API into component with section components](https://github.com/Automattic/woocommerce.com/issues/17433) in this PR itself. I'll re-request your review when the PR is ready. Thanks!

Thanks Simran! Are you then ensuring both the Discover and Extensions page are populated from the API? If so, could you assign that issue to yourself and link the PR to it so we can keep track of what's being worked on etc?

simranvijwani commented:
@kdevnel The goal for this PR was only to create product cards. To achieve this, I retrieved data from the [featured products API](https://woocommerce.com/wp-json/wccom-extensions/2.0/featured) to populate product cards.

Initially, I had not considered populating the extension page since it required accessing the different API. However, upon noticing no issue was logged for populating the extension page, I have taken care of it in this PR.

Unfortunately, this made the PR lengthy and complex for the reviewer to review the code changes. Therefore, I will be splitting this PR into multiple PRs now. Let me know if this sounds good.

kdevnel commented:
Sounds good to me Simran 😃

andfinally commented:
Thanks for tackling the next stuff Simran – I'll hold off on reviewing till you've had a chance to reorganise the PRs.

simranvijwani commented:
@andfinally @kdevnel This PR is now ready for review. I've updated the testing instructions as well.

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,145 @@
+		flex-direction: column;
+		gap: $medium-gap;
+		padding: $large-gap;
+		border-radius: $grid-unit-05;
```
This is being overridden by the product card default styling and so the card only ends up with a 1px border-radius.

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,145 @@
+
+}
+
+@media screen and (max-width: $breakpoint-medium) {
```
Not a dealbreaker, totally optional, but we're mixing desktop-first and mobile-first breakpoints. Other parts of the marketplace are using min-width breakpoints to handle things mobile-first.

andfinally replied:
Ya I agree with this, I like to start with mobile first and build up using `min-width` – I find it easier to understand.

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -0,0 +1,76 @@
+						{ product.price === 0 || product.price === '0'
+							? __( 'Free download', 'woocommerce' )
+							: product.price +
+							  window.wcSettings.admin.currency.symbol }
```
Both optional: 

- there is a helper `getAdminSetting()` function to help get settings from this global variable. 
- Is it worth adding some optional chaining in case the currency symbol isn't in the API for whatever reason?

kdevnel reviewed (changes requested):
This is looking fabulous! A couple of design-related notes and some optional comments.

The `<ProductCard>` component has some default styling that isn't quite the same as the design. I notice that the card wrapper has a 24px padding, and the card content also has a 24px padding. The design only calls for 24px in total between the card wrapper and the content.

Also, I think this might have been missed in my skeleton, but there should be 48px between the bottom row of cards and the footer.

simranvijwani commented:
Thank you for the review @kdevnel! I have implemented all your suggestions.

Dan-Q commented:
> I noticed on my side that one of the cards didn't manage to load a vendor image. It only showed some text in it's place. I wonder if we should have a local placeholder that loads instead - it would look nicer and I can imagine some user sites will have weird connection timeouts that we can't always control. The joys of planning for core!

I believe the image loads directly from WooCommerce.com, right? So any timeout is a browser<->WooCommerce.com issue, not directly related to the store's configuration?

Although now I think about it, there _are_ HTTP headers a store might rightly configure that could really muck these images up, e.g. a `Content-Security-Policy: img-src 'self'` header (which says "all images used _on_ this domain will be served _from_ this domain") would instruct browsers not to load these images.

No new action needed, just an observation.

Dan-Q commented:
I'm concerned about using `overflow: hidden` to hide product titles. On narrow screens, this makes it very difficult to differentiate between similar products: look at this example using several of our real products as test data:

https://github.com/woocommerce/woocommerce/assets/53293/201a7727-522e-4d61-b131-1aa2d0550696

It feels to me like it's more important that we _show full product titles_ than that the _cards are a specific height_ (summoning @poligilad-auto for design input). If so, I'd suggest we:

### 1. Allow flexible height product cards
- Stop forcing heights (i.e. remove `height` from `&__meta` and `&__details` in `product-card.scss`),
- Stop hiding content to make things fit (i.e. remove `overflow: hidden` and `text-overflow: ellipsis` from `&__title`), and
- Position icons centrally (i.e. set `align-items: flex-start` in `&__details`)

This introduces a new problem: some product cards can be heigher than others, and at some screen widths some can sometimes be narrower than others. This makes the design look higgldy-piddldy, but we can fix that with CSS grid:

### 2. Use CSS grid to _lay out_ product cards

I started mocking this up but then I noticed that this PR specifically _removes_ CSS grid, so I assume there's a reason? My thinking would be that in `product-list-content.scss` you do something like:

- Switch `&__product-list-content` to `display: grid` with `grid-template-columns: repeat(3, 1fr);`
- Add breakpoints so that there are fewer columns on narrower screens, e.g. `@media (width <= $breakpoint-large) { grid-template-columns: repeat(2, 1fr); }`
- Optionally add a `max-width` if you don't want product cards to ever be _super_ wide.

However, without knowing for sure why you decided to _remove_ the CSS grid, I can't be certain that this is a good idea either.

What're your thoughts?

Dan-Q reviewed (changes requested):
[I still think we've got some responsive problems](https://github.com/woocommerce/woocommerce/pull/39173#issuecomment-1655539295). Sorry!

poligilad-auto commented:
I see your point @Dan-Q. Initially I wanted to truncate the title of the extension so everything aligns perfectly 🙂 but I do think it's more important to see the name.

I think we can do 2 things or combine the two, it's partially what you suggested.
1. Show 2 items per row for narrow view ports, like you suggested.
2. Checkout [how the plugins screen solves it](https://prnt.sc/rQz8PgztGuZD). The height of all the items in a row is determined by the highest item. All the content sticks to the top, while the info at the bottom always sticks to the bottom, in this way things are relatively aligned and they still manage to show all info.

simranvijwani commented:
Thank you for the suggestions, @Dan-Q and @poligilad-auto 

I like option 2. This will help us achieve a consistent appearance for all product cards. I'll try to implement it. 

@Dan-Q about the CSS grid:

>Use CSS grid to lay out product cards

I changed it to Flex as per the CSS present in the [Figma design](2sXEw33oLyTZ7Mu3808sed-fi-1018_179513).

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/extensions/extensions.scss:
```diff
@@ -1 +1,8 @@
-// To keep StyleLint happy. Remove when file contains actual code.
+.woocommerce-marketplace {
+
```
I would suggest no line break here.

```suggestion
```

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,149 @@
+@import '../../stylesheets/_variables.scss';
+
+.woocommerce-marketplace {
+
```
```suggestion
```

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,149 @@
+			padding: 0;
+
+		}
+		&__vendor span {
```
```suggestion

		&__vendor span {
```

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,149 @@
+			text-overflow: ellipsis;
+
+		}
+		&__vendor {
```
```suggestion

		&__vendor {
```

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,149 @@
+	.woocommerce-marketplace {
+		&__product-card {
+			margin-top: 0;
+			flex-basis: inherit;
+
+
+		}
```
```suggestion
			flex-basis: inherit;
		}
```

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,149 @@
+			font-weight: 400;
+			line-height: $medium-gap;
+		}
+	}
+
+
+}
```
```suggestion
	}
}
```

andfinally commented on plugins/woocommerce-admin/client/marketplace/stylesheets/_variables.scss:
```diff
@@ -5,12 +5,14 @@ $small-gap: $grid-unit-10;   // 8px
 $medium-gap: $grid-unit-20;  // 16px
 $large-gap: $grid-unit-30;   // 24px
 $xlarge-gap: $grid-unit-40;  // 32px
+$huge-gap: $grid-unit-60;    // 48px
```
Lol, I can see us graduating to `$ultra-massive-gap` at some point. Would it make sense to rename `$xlarge-gap` to `$large-gap` and making this one `$xlarge-gap`? I'm even wondering if we really need to create synonyms for `$grid-unit-20` etc – perhaps it would make sense to use them directly.

simranvijwani replied:
Yes, it doesn't make sense to have `$huge-gap`. I removed as it won't be used often in our codebase. Thank you!

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -0,0 +1,77 @@
+				<div className="woocommerce-marketplace__product-card__header">
+					<div className="woocommerce-marketplace__product-card__details">
+						{ product.icon && (
+							<div className="woocommerce-marketplace__product-card__icon">
+								<img
+									src={ product.icon }
+									alt={ product.title }
+								/>
+							</div>
```
I get the feeling we have some unnecessary elements in the markup in this component. For example, do we really need the div here? Could we make this work with just the `img`?

andfinally reviewed (commented):
Nice work! 

I'm not sure about the approach of using `overflow: hidden` to deal with the overflow problem: it seems a shame to lose some of the text on certain viewports. 

I recognise this is a tricky problem to solve. We had to deal with this also with [WCCOM product cards](https://woocommerce.com/search/?q=shipping&collections=product&page=1). There, the cards are `display: flex` with a `min-height`, but there's nothing stopping them from getting taller. We truncate that text to a certain length before it gets into the card. But if you add more text the card grows.

<img width="250" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/f26f406b-eea6-4526-ac6b-04cf2b124256">

We do use a handy trick with the product title, [line clamping](https://css-tricks.com/almanac/properties/l/line-clamp/). This truncates the title to a maximum number of lines. The element has these styles:

```css
.wccom-card__title-link {
	-webkit-box-orient: vertical;
	-webkit-line-clamp: 4;
	display: -webkit-box;
	max-height: 108px;
	overflow: hidden;
}
```

It's similar to the technique we're using here, but maybe cuts off less of the text, because it allows a certain number of lines. It may or may not be useful here! 🤔

Dan-Q commented:
> We do use a handy trick with the product title, [line clamping](https://css-tricks.com/almanac/properties/l/line-clamp/).

Note that the code sample @andfinally's provided has a couple of issues in its Flexbox implementation. It'll work, probably, but it's not recommended:

```
.wccom-card__title-link {
	-webkit-box-orient: vertical;
	-webkit-line-clamp: 4;
	display: -webkit-box;
	max-height: 108px;
	overflow: hidden;
}
```

`-webkit-box-orient` is a webkit-specific variant of `box-orient`, which is [a non-standard _and deprecated_](https://developer.mozilla.org/en-US/docs/Web/CSS/box-orient) way of doing what would later be done better by Flexbox (which is now universally supported). `display: -webkit-box` is part of the same standard.

Instead of `display: -webkit-box;`, say `display: flex;`, and instead of `-webkit-box-orient: vertical;`, say `flex-direction: column`. It does the same thing, works in more browsers, and isn't at risk of being removed from existing browsers or breaking in weird ways.

Note that `-webkit-line-clamp:` _doesn't_ suffer from the same problem because it's become the de facto design for what will one day be `line-clamp:`. Frankly I'm amazed this one still needs a prefix given its widespread support, even among non-webkit browsers!

andfinally commented:
Awesome, thanks for the pointers Dan! 👌

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.scss:
Just noting that this PR removes the breakpoints that display 4 cards on wider screens. You can see this here - 2sXEw33oLyTZ7Mu3808sed-fi-1305_198556

kdevnel reviewed (changes requested):
Just a comment regarding a missing breakpoint that I noticed while reviewing https://github.com/woocommerce/woocommerce/pull/39465/

Dan-Q commented:
Poli provided some guidance on the questions we've all raised above at Fire Sync this week: give any Fire person a shout after you're back from AFK if the notes aren't sufficient.

simranvijwani commented on plugins/woocommerce-admin/client/marketplace/stylesheets/_variables.scss:
```diff
@@ -11,12 +11,14 @@ $content-spacing-small: $grid-unit-20;
 $breakpoint-medium: 769px;
 $breakpoint-large: 1024px;
 $breakpoint-xlarge: 1500px;
+$breakpoint-huge: 1920px;
```
Using the [Gutenberg xhuge breakpoint](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_breakpoints.scss#L6) value to display 4 product cards in a row on wider screens.

simranvijwani commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -0,0 +1,120 @@
+	const currencySymbol = getAdminSetting( 'currency' )?.symbol;
+
+	// Adjust the height of all product cards to match the tallest one.
+	const adjustProductCardHeights = () => {
```
Even with `display: grid`. Product cards were getting displayed with different heights. Added this function to adjust the height of product cards.

andfinally replied:
This strikes me as a bodge, and it feels off that we're using DOM methods inside a React component. I do see that without it, we get rows of uneven height.

<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/2d919b62-0682-414b-9019-32b5689544bf">

We can achieve something like the same result if we add `grid-auto-rows: 1f` to the container.

```diff
Index: plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss b/plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss
--- a/plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss	(revision 95d1639c59ef3d6ba761d88496364fc0eb2b63bc)
+++ b/plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss	(date 1691148269370)
@@ -12,6 +12,8 @@
 			align-items: flex-start;
 			gap: $medium-gap;
 			justify-content: space-between;
+			grid-auto-rows: 1fr;
+			height: 100%;
 		}
 
 		&__header {

```

<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/490dd0d8-de18-498e-84fa-09cd087bfbd5">

This does give rise to some excessive vertical white space inside the cards. Not sure what's causing this. Maybe the row height is ignoring the line clamping.

andfinally replied:
We can reduce the white space inside the cards if we remove the `gap: 16px;` from `.woocommerce-marketplace__product-card__content`.

<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/f93b167c-782a-4889-8b59-f17a6755f201">

simranvijwani commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,111 @@
+		}
+
+		&__title {
+			display: -webkit-box;
```
>Instead of display: -webkit-box;, say display: flex;, and instead of -webkit-box-orient: vertical;, say flex-direction: column. It does the same thing, works in more browsers, and isn't at risk of being removed from existing browsers or breaking in weird ways.

I tried changing it to `display: flex; flex-direction: column`, but the line clamp doesn't seem to work with Flex. Reverted it to `display: -webkit-box;`

andfinally replied:
Interesting, so my original example was right! It does [look like](https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-line-clamp) `flex` doesn't support it yet.

simranvijwani commented:
Thank you for the inputs,  @andfinally, @Dan-Q and @kdevnel!

I have now changed the layout to Grid and implemented all the suggestions in 95d1639c59ef3d6ba761d88496364fc0eb2b63bc.

kdevnel reviewed (approved):
Thanks for the changes. Looks good from what I can see!

andfinally commented on plugins/woocommerce-admin/client/payments-welcome/index.tsx:
```diff
@@ -33,6 +33,9 @@ declare global {
 					cta_label: string;
 					tc_url: string;
 				};
+				currency: {
+					symbol: string;
+				};
```
This seems like an unrelated change – presumably something from trunk?

simranvijwani replied:
@andfinally We were getting a TS error related to the currency symbol we were using from `WCSettings` in Product Card component. To resolve this, I added the type for it in its definition.

andfinally reviewed (commented):
Thanks for switching to CSS grid, the UI is looking really nice! 👌 

In the current marketplace, the product title links to the product page, as well as the View details/download button.

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,111 @@
+		&__details {
+			display: flex;
+			justify-content: flex-start;
+			align-items: center;
```
To prevent the situation where icons don't line up:

<img width="1576" alt="image" src="https://github.com/woocommerce/woocommerce/assets/53293/94e0d17d-8297-4b20-a2e4-1289be90eeb8">

How about: 

```suggestion
			align-items: flex-start;
```

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,111 @@
+		}
+
+		&__header {
+			grid-row: auto;
+			grid-column: 1;
```
Don't think these lines are doing anything (`grid-row: auto` is the default value, and `grid-column: 1` is implied by the fact that there's only one column).

```suggestion
```

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,111 @@
+			display: -webkit-box;
+			-webkit-box-orient: vertical;
+			-webkit-line-clamp: 3;
+			grid-row: auto;
+			grid-column: 1;
```
Don't think these lines are doing anything (grid-row: auto is the default value, and grid-column: 1 is implied by the fact that there's only one column).

```suggestion
```

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -0,0 +1,120 @@
+	const { product } = props;
+	const currencySymbol = getAdminSetting( 'currency' )?.symbol;
+
+	// Adjust the height of all product cards to match the tallest one.
+	const adjustProductCardHeights = () => {
+		const productCards = document.querySelectorAll(
+			'.woocommerce-marketplace__product-card'
+		);
+
+		let maxContentHeight = 0;
+
+		productCards.forEach( ( card ) => {
+			const content = card.querySelector(
+				'.woocommerce-marketplace__product-card__content'
+			) as HTMLElement;
+			if ( content ) {
+				maxContentHeight = Math.max(
+					maxContentHeight,
+					content.offsetHeight
+				);
+			}
+		} );
+
+		productCards.forEach( ( card ) => {
+			const container = card.querySelector(
+				'.woocommerce-marketplace__product-card__content'
+			) as HTMLElement;
+			if ( container ) {
+				container.style.height = maxContentHeight + 'px';
+			}
+		} );
+	};
+
```
Drop all of this. We can do it in CSS!

```suggestion
```

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,111 @@
+
+		&__content {
+			display: grid;
+			grid-template-columns: 1fr;
```
This does nothing. Delete.

```suggestion
```

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -0,0 +1,111 @@
+			align-items: flex-start;
+			gap: $medium-gap;
+			justify-content: space-between;
+		}
```
```suggestion
				height: 100%;
				grid-template-rows: auto 1fr 20px;
		}
```

This replaces all the JS I've suggested removing. `height: 100%` ensures the grid "fills" the card. Then the `grid-template-rows` with the `1fr` for the description area makes sure that the bottom links stay at the bottom.

Alternatively, use `grid-template-rows: 70px 1fr 20px;` to ensure that the descriptions all line up (at the expense of some extra whitespace above descriptions of products with one-line titles).

Dan-Q reviewed (changes requested):
How about this as a way to ditch the JS and fix the layout issues?

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -0,0 +1,120 @@
+					<span>
+						{ product.price === 0 || product.price === '0'
+							? __( 'Free download', 'woocommerce' )
+							: product.price + currencySymbol }
```
The design has the currency symbol first.

```suggestion
							: currencySymbol + product.price }
```

simranvijwani replied:
Thank you, I missed updating as per the new design. [Old design](2sXEw33oLyTZ7Mu3808sed-fi-1018_179513) has it at the end.

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -0,0 +1,120 @@
+		return () => {
+			window.removeEventListener( 'resize', adjustProductCardHeights );
+		};
+	}, [] );
```
We would also drop this.

andfinally reviewed (commented):
Fabulous, thanks for the tips @Dan-Q! 🙌 Here are Dan's suggestions as a patch, as all the inline comments may be confusing.

[product-card.patch.txt](https://github.com/woocommerce/woocommerce/files/12261156/product-card.patch.txt)

I've also included two optional other changes – I removed the line clamping from the product title, as it's important, and now works OK at any length, and I've raised it by 4px to line up properly with the top of the icon. (The browser takes the font line height into account, creating a visual space.)

With these tweaks I'd say we'll be good to go!

<img width="500" src="https://github.com/woocommerce/woocommerce/assets/1647564/8dccff24-cf89-4be7-8abb-d9b7780bfd17">

poligilad-auto commented:
Just wanted to clarify a comment I made before.
In this example, in the 2nd row, we could have the height of all the product cards in that row **be determined by the highest card _in that row_**. That is how the Plugins page solves it currently.

Can we do that?

<img width="1757" alt="image" src="https://github.com/woocommerce/woocommerce/assets/83961704/c2069473-2345-4999-92b8-b42416cbbdae">

andfinally commented:
> we could have the height of all the product cards in that row be determined by the highest card in that row

@poligilad-auto This is not really possible with CSS grid – I believe it's all cards, not by row.

poligilad-auto commented:
> @poligilad-auto This is not really possible with CSS grid – I believe it's all cards, not by row.

What are the advantages of a CSS grid then? Imagine have 70 search results and them all being super high and taking up space just because of one product card.

I'm starting to think maybe we need a solution of CSS grid + some truncation. Maybe we allow for max 2 lines for the product name. That's more than enough. WDYT?

andfinally commented:
> Maybe we allow for max 2 lines for the product name. That's more than enough. WDYT?

That's what Simran had originally – I'm sure it'll work for most names. I'm cool with that, especially if it helps us avoid having to make a lot of CSS changes. Independent row heights did look a bit untidy when I was fooling around in the browser. We can always come back and do this better later, if it's no good.

simranvijwani commented:
Thank you @poligilad-auto @andfinally @Dan-Q for the inputs. 

I have added line-clamp again to allow for max 2 lines for the product names. 

Here is how it looks now:

<img width="1155" alt="image" src="https://github.com/woocommerce/woocommerce/assets/34896144/adbd1a9a-4f3d-489e-809d-4baf1301de6e">

I think it's fine to limit product names to two lines since there likely won't be many products with longer names.

kdevnel reviewed (approved):
Looks good! Approving the changes as they are. I'd just say the final thing is rebasing this and ensuring everything still works as intended.

Dan-Q reviewed (approved):
I agree with Kyle. This seems to work, and anything that we discover that needs changing after merging can be fixed after merging: let's roll!

andfinally commented:
I'm finding we show `By` when there's no vendor – can we please make sure we just skip it in this case, like we do in the current marketplace? Hopefully we can squeeze this into this PR.

<img width="373" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/c78ed00f-97d8-4888-aedc-dda3873b7514">

andfinally committed:
Changes from `add/marketplace-product-card-component` after rebasing on feature branch.

andfinally commented:
Added a couple of small changes.

- Fixed API URL in `fetchCategories` function of `CategorySelector`.
- Added logic to `ProductCard` to not render vendor if vendor name isn't present.
- Changed `fetchDiscoverPageData` to use API hostname from `constants.ts`.
